sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -4003,6 +4003,56 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
     }
     checkAnswer(sql(s"select /*+ REPARTITION(3, a) */ a b from values('123') t(a)"), Row("123"))
   }
+
+  test("SPARK-35630: ExpandExec should not introduce unnecessary exchanges") {
+    withTable("test_table") {
+      spark.range(11)
+        .withColumn("group1", $"id" % 2)
+        .withColumn("group2", $"id" % 4)
+        .withColumn("a", $"id" % 3)
+        .withColumn("b", $"id" % 6)
+        .write.saveAsTable("test_table")
```

Comment:
nit: for better test performance, could you use a temporary view instead?

Reply:
I simplified it even a bit further